src/_zkapauthorizer/backup-recovery.yaml

```diff
@@ -0,0 +1,136 @@
+openapi: "3.0.0"
+
+info:
+  title: "Replication / Recovery"
+  description: >-
+    This API allows replication and recovery of ZKAPAuthorizer internal state.
+    Replication is in a single direction from this single Tahoe-LAFS node to
+    the Tahoe-LAFS grid.
+  version: "1.0.0"
+
+paths:
+  /storage-plugins/privatestorageio-zkapauthz-v1/recover:
+    post:
+      description: >-
+        Restore from a previously configured replica.
+
+        This is only valid when ZKAPAuthorizer has no local state yet.
+      requestBody:
+        content:
+          application/json:
+            schema:
+              "$ref": "/components/schemas/ReplicaConfiguration"
+
+      responses:
+        "500": # INTERNAL SERVER ERROR
+          description: >-
+            Recovery from the replica has failed for some reason.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Error"
+
+        "409": # CONFLICT
+          description: >-
+            Recovery is not possible at this time because local state already
+            exists in the database.  Recover into a new, empty ZKAPAuthorizer
+            installation only.
+
+        "404": # NOT FOUND
+          description: >-
+            Recovery from the replica has failed because the recovery data
+            cannot be retrieved using the given capability.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Error"
+
+        "200": # OK
+          description: >-
+            Recovery from the replica has succeeded.
+          content:
+            application/json:
+              schema:
+                type: "object"
+                properties: {}
+
+  /storage-plugins/privatestorageio-zkapauthz-v1/replicate:
+    get:
+      description: >-
+        Get this node's replica configuration.
+      responses:
+        404: # NOT FOUND
+          description: >-
+            This node is not configuration for replication.
+
+        200: # OK
+          description: >-
+            Information about replica configuration is available and included
+            in the response.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/ReplicaConfiguration"
+
+    post:
+      description: >-
+        Configure an on-grid replica.
+
+        Once a replica has been configured this node will keep that replica
+        up-to-date with local state changes.
```

Comment:
It might be useful for callers to be able to manually/force a replica to be created/uploaded on demand (rather than needing to wait until that object is of "large enough" size -- for example, in cases where a user has just finished uploading all of their files and wants to be sure that their remote database is fully up-to-date before they turn off their laptop.

Reply:
Perhaps an "even better" use-case would be a user preparing to go offline for a while (or cross a border) wanting to ensure stuff is up-to-date.

Or, a user may be _planning_ to recover later (e.g. they're going to wipe their device, maybe they're crossing a border or giving it back to an employer or to a repair shop).

Reply:
Maybe it should synchronize the replica before allowing the Tahoe-LAFS client node process to exit?  Then every time you cleanly quit the app the replica is up to date.

Reply:
I know this was discussed elsewhere, and I certainly agree that could be a good Tahoe feature .. but I think it's "complex" to get correct there, even with only a single plugin (e.g. zkap-authorizer needs to be shut down _first_ and ~all the other Tahoe services need to be alive while it does so).

So, both pragmatically and under the "explicit is better than implicit" banner, an endpoint to ask zkapauthorizer in particular to do this would be good. (However, under the "it's complex" part, then "something" -- zkap or gridsync -- needs to wait until Tahoe is done).

Perhaps a hybrid solution: Tahoe grows a feature where a "nice" exit signal means it finishes all "active operations" _and_ zkapauthorizer has an explicit endpoint to "backup all outstanding data, now" ... so then gridsync calls the zkap-authorizer endpoint (queuing up any required active Tahoe operations), then kills tahoe, then waits for Tahoe to exit.

As another "plus" for an explicit endpoint: it may be that a user _doesn't_ care about the data being in the grid (e.g. a "normal close of the app") and just wants a quick but orderly shutdown. (I'm not trying to actually design the UX here, but it's a lot more flexible and "up to the UI" with the endpoint versus implicit stuff like "I hope that TERM means my data got there").

Reply:
I added a note about the "safe border crossing" use-case to the roadmap doc.